cocotb/handle.py

```diff
@@ -530,11 +519,67 @@ def _range_iter(self, left, right):
 
     @NonHierarchyObject.value.getter
     def value(self):
-        # need to iterate over the sub-object
-        result = []
-        for x in range(len(self)):
-            result.append(self[x].value)
-        return result
+        """Reference to the value of the simulation object.
+
+        Getting and setting the current value of an array is done
+        by iterating through sub-handles in left-to-right order.
+
+        **Getting value**:
+
+        The returned list contains the values of each sub-handle
+        in the order indicated in the table:
+
+        +--------------+---------------------+--------------------------------------+
+        | Verilog      | VHDL                | Python list                          |
+        +==============+=====================+======================================+
+        | ``arr[4:7]`` | ``arr(4 to 7)``     | ``[arr[4], arr[5], arr[6], arr[7]]`` |
+        +--------------+---------------------+--------------------------------------+
+        | ``arr[7:4]`` | ``arr(7 downto 4)`` | ``[arr[7], arr[6], arr[5], arr[4]]`` |
+        +--------------+---------------------+--------------------------------------+
+
+        **Setting value**:
+
+        The indices in the list ``value`` will be assigned to sub-handles
+        in the order indicated in the table:
+
+        +--------------+---------------------+--------------------------------+
+        | Verilog      | VHDL                | Array values                   |
+        +==============+=====================+================================+
+        | ``arr[4:7]`` | ``arr(4 to 7)``     | - ``arr[4]`` gets ``value[0]`` |
+        |              |                     | - ...                          |
+        |              |                     | - ``arr[7]`` gets ``value[3]`` |
+        +--------------+---------------------+--------------------------------+
+        | ``arr[7:4]`` | ``arr(7 downto 4)`` | - ``arr[7]`` gets ``value[0]`` |
+        |              |                     | - ...                          |
+        |              |                     | - ``arr[7]`` gets ``value[3]`` |
+        +--------------+---------------------+--------------------------------+
```

Comment:
I'd argue we can combine this table with the one above - the table above already contains the information, namely that
```python
val = arr.value
# means
val = [arr[4].value, arr[5].value, arr[6].value, arr[7].value]
```
and
```python
arr.value = val
# means
[arr[4].value, arr[5].value, arr[6].value, arr[7].value] = val
```

Perhaps the table becomes
```
        +--------------+---------------------+--------------------------------------------------------------+
        | Verilog      | VHDL                | ``arr.value`` is equivalent to                               |
        +==============+=====================+==============================================================+
        | ``arr[4:6]`` | ``arr(4 to 7)``     | ``[arr[4].value, arr[5].value, arr[6].value, arr[7].value]`` |
        +--------------+---------------------+--------------------------------------------------------------+
        | ``arr[7:4]`` | ``arr(7 downto 4)`` | ``[arr[7].value, arr[6].value, arr[5].value, arr[4].value]`` |
        +--------------+---------------------+--------------------------------------------------------------+

        Note that this property can be set, and the same equivalence holds in ``arr.value = ...``.
```

Not entirely sure how to word that last sentence.

Reply:
That's pretty slick. Thanks for the feedback on this, good documentation can be hard.

Reply:
It's a lot easier when the contributor writes such detailed explanations of the intent in the PR comments, as you've done above :)